inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gemm_mmad_int8_slm.cl

```diff
@@ -0,0 +1,229 @@
+
+    // Batch offsets
+    const uint batch_offset_input0 = FUNC_CALL(get_input0_batch_offset)(b, f, w, z);
+    const uint batch_offset_input1 = FUNC_CALL(get_input1_batch_offset)(b, f, w, z);
+#ifdef INPUT2_TYPE
+    const uint batch_offset_input2 = FUNC_CALL(get_input2_batch_offset)(b, f, w, z);
+#endif // INPUT2_TYPE
+    const uint batch_offset_output = FUNC_CALL(get_output_batch_offset)(b, f, w, z);
+
+    // Pointer for loading the matrix B from the global memory
+    __global PACKED_INPUT1_TYPE_VEC* input1_pnt = (__global PACKED_INPUT1_TYPE_VEC*)input1;
+
+    // SLM tile of the matrix B
+    __local PACKED_INPUT1_TYPE_VEC slm_tile_input1[SLM_TILE_SIZE * SLM_DECIMATION_FACTOR];
+
+    // Pointer for loading the matrix B from SLM to registry chunks (GRF)
+    __local INPUT1_TYPE* slm_tile_input1_pnt = (__local INPUT1_TYPE*)slm_tile_input1;
+
+    // Registry chunks of input matrices (A, B) + input2 (optional)
+    PACKED_INPUT0_TYPE_VEC reg_tile_input0;
+    PACKED_INPUT1_TYPE_VEC reg_tile_input1;
+#ifdef INPUT2_TYPE
+    ACTIVATION_TYPE_VEC tile_input2;
+#endif // INPUT2_TYPE
+
+    // Registry chunks of the output matrix (C)
+    ACCUMULATOR_TYPE_VEC reg_tile_output[4] = { (ACCUMULATOR_TYPE_VEC)(ACCUMULATOR_VAL_ZERO),
+                                                (ACCUMULATOR_TYPE_VEC)(ACCUMULATOR_VAL_ZERO),
+                                                (ACCUMULATOR_TYPE_VEC)(ACCUMULATOR_VAL_ZERO),
+                                                (ACCUMULATOR_TYPE_VEC)(ACCUMULATOR_VAL_ZERO) };
+
+    // Pointer to the result array (the matrix C)
+    ACCUMULATOR_TYPE* reg_tile_output_pnt = (ACCUMULATOR_TYPE*)reg_tile_output;
+
+    // Calculating positions for loading input matrices from the global memory
+    const uint wg_offset_y = lid1 * SUB_GROUP_SIZE + lid0;
+    const uint input1_index = (batch_offset_input1 + wg_offset_y * INPUT1_SIZE_X) / SLM_TILE_SIZE + gid0;
+    const uint common_input0_offset = batch_offset_input0 + (gid1 * SLM_TILE_SIZE + lid1 * SUB_GROUP_SIZE) * INPUT0_SIZE_X;
+
+#ifdef PRELOADING_SLM
+    for (uint i = 0; i < SLM_DECIMATION_FACTOR; i++) {
+        slm_tile_input1[i * SLM_TILE_SIZE + wg_offset_y] = input1_pnt[input1_index + i * INPUT1_SIZE_X];
+    }
+
+    // Synchronization; waiting until all work items will finish loading the matrix B from the global memory to SLM
+    barrier(CLK_LOCAL_MEM_FENCE);
+#endif // PRELOADING_SLM
+
+    // Loop by "k" tiles
+    for (uint k = 0; k < INPUT0_SIZE_X / SLM_TILE_SIZE; k++) {
+
+        // Loading the matrix A from the global memory to GRF
+        for (uint i = 0; i < SUB_GROUP_SIZE; i++) {
+            reg_tile_input0[i] = AS_TYPE(PACKED_INPUT0_TYPE, BLOCK_READ(input0 + common_input0_offset + k * SLM_TILE_SIZE + i * INPUT0_SIZE_X));
+        }
+
+#ifndef PRELOADING_SLM
+        // Loading the matrix B to SLM
+        if (k % SLM_DECIMATION_FACTOR == 0) {
+
+            // Synchronization
+            barrier(CLK_LOCAL_MEM_FENCE);
```

Comment:
Why this synchronization is needed?

Reply:
Because without this synchronization we'll get incorrect result.